### PR TITLE
#1361 v3 basic send logic (part 1)

### PR DESCRIPTION
**Pull request by kalbfled**
# Description

These changes for v3 notification logic:
- Implement a 501 (not implemented) response if a user includes "scheduled_for" in the request data
- Include Celery tasks to send notifications with Pinpoint (SMS) and SES (e-mail)
- Include unit tests for the above

The changes are nowhere near feature complete, and this PR does not complete #1361.  I will author a new PR with additional changes that add error handling, which is necessary, among other reasons, to retry failed tasks when appropriate.

Many other tickets are necessary before v3 achieves anything resembling feature completeness or parity with v2.  There are a lot of TODOs in the code, and creating tickets for all of them now is not feasible.  Please trust that I'm working on all of it.

issue #1361

## Type of change

Please check the relevant option(s).

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Documentation changes only

## How Has This Been Tested?

In addition to new unit tests, I have successfully sent myself SMS and e-mail notifications using Postman.  Please see the ticket comments for specific Dev notification IDs you can GET, or try Postman yourself.  (Make sure you have imported the latest Postman collection.)

For SMS, you must provide a valid sms_sender_id right now.  You won't see personalizations either; just the raw template.

## Checklist

- [x] I have assigned myself to this PR
- [x] PR has an appropriate title: #9999 - What the thing does
- [x] PR has a detailed description, including links to specific documentation
- [x] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/notification-api/blob/master/.github/release.yaml) to the PR.
- [x] I did not remove any parts of the template, such as checkboxes even if they are not used
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to any documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have ensured the latest master is merged into my branch and all checks are green prior to review
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] The ticket is now moved into the DEV test column

### CODE REVIEW

File: tests/app/celery/v3/test_notification_tasks.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    assert sample_template_without_email_permission.template_type == EMAIL_TYPE
+    assert sample_template_without_email_permission.service_id != sample_service.id
+    assert False
+
+    request_data = {
+        "id": str(uuid4()),
+        "notification_type": SMS_TYPE,
+        "phone_number": "+18006982411",
+        "template_id": sample_template.id,
+    }
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply tests/app/celery/v3/test_notification_tasks.py:68]: I'm aware of this.  It's in a test that is xfail'ed right now.

[the remark this answers is not in the document]
Comment (cris-oddball) [reply tests/app/celery/v3/test_notification_tasks.py:68]: To make the CodeQL overlords happy, maybe just comment out the entire test for now?

File: tests/app/celery/v3/test_notification_tasks.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    assert notification.status == NOTIFICATION_PERMANENT_FAILURE
+    assert notification.status_reason == "The template type does not match the notification type."
+
+
+def test_v3_process_notification_template_type_mismatch_2(
+    notify_db_session, mocker, sample_service, sample_email_template
+):
+    """
+    Call the task with request data for an SMS notification, but specify an e-mail template.
+    """
```
Comment (cris-oddball) [inline tests/app/celery/v3/test_notification_tasks.py:109]: Is this an effective unit test, given that integrated, sending an SMS with an email template fails?

Comment (kalbfled) [reply tests/app/celery/v3/test_notification_tasks.py:109]: It's effective in that it tests the logic I wrote.  The failure, which we have captured with a spike ticket, is outside of the v3 code.

File: tests/app/celery/v3/test_notification_tasks.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kalbfled) [reply tests/app/celery/v3/test_notification_tasks.py:67]: I removed the preceding "assert False" statement, which is what was making CodeQL complain about unreachable code.  It was in there for debugging (Pytest dumps output only when failures occur).

**Review by cris-oddball — APPROVED** [review]
Good enough for now.

File: app/v3/notifications/rest.py [hunk 3]
```diff
@@ -96,5 +111,13 @@ def v3_send_notification(request_data: dict) -> str:
     # This has the side effect of modifying the input in the upstream code.
     request_data["id"] = str(uuid4())
 
-    # TODO 1361 - Pass the validated request data to Celery by calling apply_async (or something else)
+    # Initiate a Celery task to process the validated request data.  This does not block.
+    # TODO - v2 uses the imported value "api_user" for the api_key.
```
Comment (EvanParish) [inline app/v3/notifications/rest.py:115]: Are all of the TODOs spread throughout this code for the part 2 PR, or are they associated with follow-up tickets?

Comment (kalbfled) [reply app/v3/notifications/rest.py:115]: Both, and I haven't written the follow-up tickets.  This ticket (#1361) has a list of tickets for additional work we know we need.  Part 2 of this ticket will handle error and retries for the Celery tasks.

Comment (cris-oddball) [reply app/v3/notifications/rest.py:115]: @EvanParish given that this is a WIP with a plan, I'm OK with the TODOs for now. On the final PR, I expect there to be no TODOs or -- if there are TODOs -- they contain ticket numbers.

**Review by EvanParish — APPROVED** [review]

**Pull request merged**